hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/DefaultOperationQuota.java

```diff
@@ -131,7 +131,8 @@ public void close() {
 
   @Override
   public long getReadAvailable() {
-    return readAvailable;
+    // at this point we've grabbed some quota, so we should use at least that
+    return Math.max(readAvailable, readConsumed);
```

Comment:
Without this change we've experienced erroneous MultiActionResultTooLarge when riding close to quotas. As `readAvailable` can be negative, this can even affect otherwise irrelevant operation types. For example, riding close to a read quota and entering negative read availability would cause multiPuts, which do no reading, to throw MultiActionResultTooLarge. The logic for throwing a MultiActionResultTooLarge can be found [here](https://github.com/HubSpot/hbase/blob/hubspot-2.5/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java#L780-L799). Worse, this exception is retried immediately, so this can snowball into a DDOS of your RPC layer.

Reply:
Should we move this Math.max to caller? Theoretically readAvailable could be negative, the caller should decide how to deal with this.

Reply:
Yeah, I wonder whether it's worth having a separate method, `OperationQuota#getMaxResultSize`, which encapsulates this logic and is not hiding business logic in what appears to be a simple getter

Reply:
I think we should keep this here. I have three reasons:

1. There are only 2 callers to this method, and both would need the Math.max() added
2. We don't currently expose readConsumed
3. Logically, I think it makes sense to consider readAvailable in this way for all cases. I could even see it being `Math.max(readAvailable + readConsumed, readConsumed)`. This is because of the nuanced quota logic in checkQuota. Details on this one below:

In checkQuota, it iterates all limiters twice. Once to check for available (or throw throttle exception). Once it gets through all limiters, it loops them again to grab the quota. This loop can over consume the quota. At this point we've grabbed the quota and there's no easy way to return it. So we might as well consider this as "available" for us.

The quotas are complicated because there can be many limiters and checkQuota is not synchronized. It'd be more accurate to synchronize the entire checkQuota, but I think it's better for performance to let it be.

Reply:
Then better rename the method, or at least add more comments to describe the logic.

Reply:
Ah sorry Bryan, I didn't see your comment until I pushed a change. I think the only difference though would be to remove the readAvailable and readConsumed getters, and move the getMaxResultSize logic out of the interface. Do you think that's worth doing, or do you like the changeset in its current form?

Reply:
Personally I prefer the original name, which has the javadoc:
>/** Returns the number of bytes available to read to avoid exceeding the quota *

To me "read available" makes sense to be "how much did we grab, plus how much is left". But I can see that might be subjective. So I don't have a strong opinion.

Reply:
I see both sides. I think it's ok for us to be more specific with the naming here because, like you said, the current usage scope is small enough for us to be specific